server/handlers/project_command_output_handler.go

```diff
@@ -172,40 +184,65 @@ func (p *AsyncProjectCommandOutputHandler) writeLogLine(pull string, line string
-	if p.projectOutputBuffers[pull] == nil {
-		p.projectOutputBuffers[pull] = []string{}
+	if p.projectOutputBuffers[jobID] == nil {
+		p.projectOutputBuffers[jobID] = []string{}
 	}
-	p.projectOutputBuffers[pull] = append(p.projectOutputBuffers[pull], line)
+	p.projectOutputBuffers[jobID] = append(p.projectOutputBuffers[jobID], line)
 	p.projectOutputBuffersLock.Unlock()
 }
 
 //Remove channel, so client no longer receives Terraform output
-func (p *AsyncProjectCommandOutputHandler) Deregister(pull string, ch chan string) {
-	p.logger.Debug("Removing channel for %s", pull)
+func (p *AsyncProjectCommandOutputHandler) Deregister(jobID string, ch chan string) {
+	p.logger.Debug("Removing channel for %s", jobID)
 	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers[pull], ch)
+	delete(p.receiverBuffers[jobID], ch)
 	p.receiverBuffersLock.Unlock()
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(pull string) map[chan string]bool {
-	return p.receiverBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetReceiverBufferForPull(jobID string) map[chan string]bool {
+	return p.receiverBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(pull string) []string {
-	return p.projectOutputBuffers[pull]
+func (p *AsyncProjectCommandOutputHandler) GetProjectOutputBuffer(jobID string) []string {
+	return p.projectOutputBuffers[jobID]
 }
 
-func (p *AsyncProjectCommandOutputHandler) CleanUp(pull string) {
-	p.projectOutputBuffersLock.Lock()
-	delete(p.projectOutputBuffers, pull)
-	p.projectOutputBuffersLock.Unlock()
+func (p *AsyncProjectCommandOutputHandler) CleanUp(jobContext models.JobContext) {
 
-	// Only delete the pull record from receiver buffers.
-	// WS channel will be closed when the user closes the browser tab
-	// in closeHanlder().
-	p.receiverBuffersLock.Lock()
-	delete(p.receiverBuffers, pull)
-	p.receiverBuffersLock.Unlock()
+	// Retrieve the jobID from the pullToJobMapping
+	if jobID, ok := p.pullToJobMapping[jobContext]; ok {
+		p.projectOutputBuffersLock.Lock()
+		delete(p.projectOutputBuffers, jobID)
+		p.projectOutputBuffersLock.Unlock()
+
+		// Only delete the pull record from receiver buffers.
+		// WS channel will be closed when the user closes the browser tab
+		// in closeHanlder().
+		p.receiverBuffersLock.Lock()
+		delete(p.receiverBuffers, jobID)
+		p.receiverBuffersLock.Unlock()
+	}
+}
+
+func (p *AsyncProjectCommandOutputHandler) GenerateJobID(pull models.PullRequest, projectName string, workspace string) string {
+	// Check if the job context exists in the pullToJobMapping
+
+	jobContext := models.JobContext{
+		PullNum:     pull.Num,
+		Repo:        pull.BaseRepo.Name,
+		ProjectName: projectName,
+		Workspace:   workspace,
+		HeadCommit:  pull.HeadCommit,
+	}
+
+	var jobID string
+	if val, ok := p.pullToJobMapping[jobContext]; ok {
+		jobID = val
+	} else {
+		jobID = uuid.New().String()
+		p.pullToJobMapping[jobContext] = jobID
```

Comment:
I think we should always be generating a new job id here.  This defeats the purpose of using uuid vs. just using a hash, you're effectively doing the same thing here. Then after generating the new UUID you can persist it in the map.

Reply:
Generating new JobIDs for every new operation as discussed during our stand up!